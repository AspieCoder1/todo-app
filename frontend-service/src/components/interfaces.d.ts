import { Document } from 'mongoose';

interface ITodo extends Document {
	body: string;
	completed: boolean;
}

interface IUser extends Document {
	email: string;
	password: string;
}
