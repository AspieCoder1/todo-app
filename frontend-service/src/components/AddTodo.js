import React, { Component } from 'react';
import classNames from 'classnames';

import styles from '../styles/components/addTodo.module.scss';
export default class AddTodo extends Component {
	render() {
		return (
			<form onSubmit={this.props.onSubmit}>
				<div className={styles.inputGroup}>
					<input
						className={classNames(styles.input, {
							[styles.inputError]:
								this.props.errors && this.props.errors.noTodo,
						})}
						type="text"
						name="todoToAdd"
						value={this.props.todoToAdd}
						onChange={this.props.onChange}
						placeholder="Enter your new todo here..."
					/>
					<button className={styles.button} type="submit">
						Add Todo
					</button>
				</div>
				{this.props.errors && this.props.errors.noTodo ? (
					<p className={styles.error}>Please provide a todo</p>
				) : null}
			</form>
		);
	}
}
