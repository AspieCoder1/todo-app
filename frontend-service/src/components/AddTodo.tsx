import * as React from 'react';
import classNames from 'classnames';
import * as styles from '../styles/components/addTodo.module.scss';

interface IErrors {
	[key: string]: string;
}

interface IProps {
	errors: IErrors | undefined;
	todoToAdd: string;
	onSubmit: (e: React.FormEvent<HTMLFormElement>) => void;
	onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const AddTodo: React.FC<IProps> = props => (
	<form onSubmit={props.onSubmit}>
		<div className={styles.inputGroup}>
			<input
				className={classNames(styles.input, {
					[styles.inputError]: props.errors && props.errors.noTodo,
				})}
				type="text"
				name="todoToAdd"
				value={props.todoToAdd}
				onChange={props.onChange}
				placeholder="Enter your new todo here..."
			/>
			<button className={styles.button} type="submit">
				Add Todo
			</button>
		</div>
		{props.errors && props.errors.noTodo ? (
			<p className={styles.error}>Please provide a todo</p>
		) : null}
	</form>
);

export default AddTodo;
