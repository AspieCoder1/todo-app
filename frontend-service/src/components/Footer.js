import React from 'react';
import moment from 'moment';

export default () => {
	return (
		<div className="footer">
			<h3>&copy;AspieCoder {moment().year()}, all rights reserved</h3>
		</div>
	);
};
