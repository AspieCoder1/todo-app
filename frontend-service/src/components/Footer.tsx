import * as React from 'react';
import moment from 'moment';

const Footer: React.FC<{}> = () => {
	return (
		<div className="footer">
			<h3>&copy;AspieCoder {moment().year()}, all rights reserved</h3>
		</div>
	);
};

export default Footer;
