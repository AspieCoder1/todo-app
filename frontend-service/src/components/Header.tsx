import React from 'react';
import headerStyles from '../styles/components/header.module.scss';

const Header: React.FC<{}> = () => (
	<div>
		<div>
			<h2 className={headerStyles.title}>
				TodoApp - Manage your Todos in peace
			</h2>
		</div>
	</div>
);

export default Header;
