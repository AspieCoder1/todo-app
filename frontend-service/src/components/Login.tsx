import * as React from 'react';
import { connect } from 'react-redux';
import classNames from 'classnames';
import { Link } from 'react-router-dom';

import { loginUser, clearErrors } from '../redux/actions/auth';
import history from '../utils/history';
import styles from '../styles/components/userForm.module.scss';
import { AxiosRequestConfig } from 'axios';
import History from 'history';

interface IErrors {
	email: string | undefined;
	password: string | undefined;
}

interface IUser {
	email: string;
	password: string;
}

interface IState {
	email: string;
	password: string;
	[key: string]: string;
}

interface IReduxProps {
	errors: IErrors;
}

interface IDispatchProps {
	loginUser: (user: IUser, history: any) => (dispatch: any) => Promise<any>;
	clearErrors: () => { type: string };
}

interface IProps extends IReduxProps {
	loginUser: (user: IUser, history: any) => (dispatch: any) => Promise<any>;
	clearErrors: () => { type: string };
}

export class Login extends React.Component<IProps, IState> {
	readonly state = {
		email: '',
		password: '',
	};

	onSubmit = (e: React.FormEvent<HTMLFormElement>) => {
		e.preventDefault();
		const user = {
			email: this.state.email,
			password: this.state.password,
		};
		this.props.loginUser(user, history);
	};

	onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		this.setState({ [e.target.name]: e.target.value });
	};

	componentWillUnmount() {
		this.props.clearErrors();
	}

	render() {
		return (
			<div className={styles.page}>
				<div className={styles.content}>
					<h1 className={styles.title}>Login</h1>
					<form className={styles.form} onSubmit={this.onSubmit}>
						<div className={styles.inputGroup}>
							<label className={styles.label} htmlFor="email">
								Email
							</label>
							<input
								className={classNames(styles.input, {
									[styles.inputError]:
										this.props.errors && this.props.errors.email,
								})}
								type="text"
								name="email"
								id="email"
								value={this.state.email}
								onChange={this.onChange}
								autoComplete="email"
							/>
							{this.props.errors && this.props.errors.email ? (
								<p className={styles.error}>{this.props.errors.email}</p>
							) : null}
						</div>
						<div className={styles.inputGroup}>
							<label className={styles.label} htmlFor="password">
								Password
							</label>
							<input
								className={classNames(styles.input, {
									[styles.inputError]:
										this.props.errors && this.props.errors.password,
								})}
								type="password"
								name="password"
								id="password"
								value={this.state.password}
								onChange={this.onChange}
								autoComplete="password"
							/>
							{this.props.errors && this.props.errors.password ? (
								<p className={styles.error}>{this.props.errors.password}</p>
							) : null}
						</div>
						<button className={styles.button} type="submit">
							Login
						</button>
					</form>
					<Link className={styles.link} to="/register">
						I need an account
					</Link>
				</div>
			</div>
		);
	}
}

// TODO: change state to match redux state interface
const mapStateToProps = (state: any): IReduxProps => ({
	errors: state.errors,
});

const mapDispatchToProps = (): IDispatchProps => ({
	loginUser,
	clearErrors,
});

export default connect(mapStateToProps, mapDispatchToProps)(Login);
