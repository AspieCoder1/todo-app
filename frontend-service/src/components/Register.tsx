import * as React from 'react';
import { Link } from 'react-router-dom';

import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import classNames from 'classnames';

import { registerUser, clearErrors } from '../redux/actions/auth';
import * as styles from '../styles/components/userForm.module.scss';

export interface IErrors {
	email: string | undefined;
	password: string | undefined;
}

interface IUser {
	email: string;
	password: string;
}

interface IState {
	email: string;
	password: string;
	[key: string]: string;
}

interface IMappedProps {
	errors: IErrors;
}

interface IDispatchProps {
	registerUser: (user: IUser) => (dispatch: any) => Promise<any>;
	clearErrors: () => { type: string };
}

interface IProps extends IMappedProps {
	registerUser: (user: IUser) => (dispatch: any) => Promise<any>;
	clearErrors: () => { type: string };
}

export class Register extends React.Component<IProps, IState> {
	readonly state = {
		email: '',
		password: '',
	};

	onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		this.setState({ [e.target.name]: e.target.value });
	};

	onSubmit = (e: React.FormEvent<HTMLFormElement>) => {
		e.preventDefault();
		const newUser = {
			email: this.state.email,
			password: this.state.password,
		};
		this.props.registerUser(newUser);
	};

	componentWillUnmount() {
		this.props.clearErrors();
	}

	render() {
		return (
			<div className={styles.page}>
				<div className={styles.content}>
					<h1 className={styles.title}>Register</h1>
					<form className={styles.form} onSubmit={this.onSubmit}>
						<div className={styles.inputGroup}>
							<label className={styles.label} htmlFor="email">
								Email
							</label>
							<input
								className={classNames(styles.input, {
									[styles.inputError]:
										this.props.errors && this.props.errors.email,
								})}
								type="text"
								name="email"
								id="email"
								value={this.state.email}
								onChange={this.onChange}
								autoComplete="email"
							/>
							{this.props.errors && this.props.errors.email ? (
								<p className={styles.error}>{this.props.errors.email}</p>
							) : null}
						</div>
						<div className={styles.inputGroup}>
							<label className={styles.label} htmlFor="password">
								Password
							</label>
							<input
								className={classNames(styles.input, {
									[styles.inputError]:
										this.props.errors && this.props.errors.password,
								})}
								type="password"
								name="password"
								id="password"
								value={this.state.password}
								onChange={this.onChange}
								autoComplete="password"
							/>
							{this.props.errors && this.props.errors.password ? (
								<p className={styles.error}>{this.props.errors.password}</p>
							) : null}
						</div>
						<button className={styles.button} type="submit">
							Register
						</button>
					</form>
					<Link className={styles.link} to="/login">
						I already have an account
					</Link>
				</div>
			</div>
		);
	}
}

const mapStateToProps = (state: any): IMappedProps => ({
	errors: state.errors,
});

const mapDispatchToProps = (dispatch: any): IDispatchProps => {
	return bindActionCreators(
		{
			registerUser,
			clearErrors
		},
		dispatch
	)
};

export default connect(mapStateToProps, mapDispatchToProps)(Register);
