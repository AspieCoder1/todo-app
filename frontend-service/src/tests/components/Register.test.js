import React from 'react';
import { shallow } from 'enzyme';
import { Register } from '../../components/Register';

it('should render register component correctly', () => {
	const wrapper = shallow(<Register />);

	expect(wrapper).toMatchSnapshot();
});

it('should change user state correctly', () => {
	const wrapper = shallow(<Register />);

	const email = 'test@test.io';
	const password = 'password';

	wrapper.find('#email').simulate('change', {
		target: {
			name: 'email',
			value: email,
		},
	});
	wrapper.find('#password').simulate('change', {
		target: {
			name: 'password',
			value: password,
		},
	});
	wrapper.update();

	expect(wrapper.state().email).toBe(email);
	expect(wrapper.state().password).toBe(password);
});

it('should handle user register correctly', () => {
	const registerUser = jest.fn();
	const wrapper = shallow(<Register registerUser={registerUser} />);

	const email = 'test@test.io';
	const password = 'password';

	wrapper.find('#email').simulate('change', {
		target: {
			name: 'email',
			value: email,
		},
	});
	wrapper.find('#password').simulate('change', {
		target: {
			name: 'password',
			value: password,
		},
	});
	wrapper.update();

	wrapper.find('form').simulate('submit', { preventDefault: () => undefined });
	expect(registerUser).toHaveBeenCalledTimes(1);
	expect(registerUser).toHaveBeenCalledWith({ email, password });
});
