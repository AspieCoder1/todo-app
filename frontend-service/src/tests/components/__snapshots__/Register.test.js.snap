// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render register component correctly 1`] = `
<div>
  <div>
    <h1>
      Register
    </h1>
    <form
      onSubmit={[Function]}
    >
      <div>
        <label
          htmlFor="email"
        >
          Email
        </label>
        <input
          autoComplete="email"
          className=""
          id="email"
          name="email"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          htmlFor="password"
        >
          Password
        </label>
        <input
          autoComplete="password"
          className=""
          id="password"
          name="password"
          onChange={[Function]}
          type="password"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Register
      </button>
    </form>
    <Link
      to="/login"
    >
      I already have an account
    </Link>
  </div>
</div>
`;
